lib_test/scheduler_test.ml

```diff
@@ -0,0 +1,116 @@
+  let x = make () in
+  let before = Unix.gettimeofday () in
+  let th = Thread.create (fun () -> wait x 0.5) () in
+  signal x;
+  Thread.join th;
+  let after = Unix.gettimeofday () in
+  let elapsed = after -. before in
+  assert_bool "elapsed_time1" (elapsed < 0.4)
+
+(* Test the injection of a one-shot function at a time in the future *)
+let test_one_shot () =
+  let after = ref None in
+  let before = Unix.gettimeofday () in
+  let _ = Scheduler.one_shot (Scheduler.Delta 1) "test_one_shot"
+      (fun () -> after := Some (Unix.gettimeofday ())) in
+  Thread.delay 2.0;
+  let success =
+    match !after with
+    | Some x ->
+      let elapsed = x -. before in
+      elapsed > 0.99 && elapsed < 2.01
+    | None ->
+      false
+  in
+  assert_bool "one_shot_success" success
+
+(* Test the injection of a one-shot function at an absolute time *)
+let test_one_shot_abs () =
+  let after = ref None in
+  let before = Unix.gettimeofday () in
+  let now = Scheduler.now () in
+  let _ = Scheduler.one_shot (Scheduler.Absolute (Int64.add 1L now)) "test_one_shot"
+      (fun () -> after := Some (Unix.gettimeofday ())) in
+  Thread.delay 2.0;
+  let success =
+    match !after with
+    | Some x ->
+      let elapsed = x -. before in
+      elapsed > 0.99 && elapsed < 2.01
+    | None ->
+      false
+  in
+  assert_bool "one_shot_success" success
+
+(* Tests that the scheduler still works even after a failure occurs in
+   the injected function *)
+let test_one_shot_failure () =
+  let after = ref None in
+  let before = Unix.gettimeofday () in
+  let _ = Scheduler.one_shot (Scheduler.Delta 0) "test_one_shot"
+      (fun () -> after := failwith "Error") in
+  let _ = Scheduler.one_shot (Scheduler.Delta 1) "test_one_shot"
+      (fun () -> after := Some (Unix.gettimeofday ())) in
+  Thread.delay 2.0;
+  let success =
+    match !after with
+    | Some x ->
+      let elapsed = x -. before in
+      elapsed > 0.99 && elapsed < 2.01
+    | None ->
+      false
+  in
+  assert_bool "one_shot_success" success
+
+(* Checks that one-shot functions can cancelled and are then not executed *)
+let test_one_shot_cancel () =
+  let after = ref None in
+  let x = Scheduler.one_shot (Scheduler.Delta 1) "test_one_shot_cancel" (fun () -> after := Some (Unix.gettimeofday ())) in
+  Scheduler.cancel x;
+  Thread.delay 2.0;
+  let success =
+    match !after with
+    | Some _ -> false
+    | None -> true
+  in
+  assert_bool "one_shot_cancelled" success
+
+(* Check that dumping the state of the scheduler contains a reference to
+   a test function that has been injected *)
+let test_dump () =
+  let after = ref None in
+  let before = Unix.gettimeofday () in
+  let _ = Scheduler.one_shot (Scheduler.Delta 1) "test_dump"
+      (fun () -> after := Some (Unix.gettimeofday ())) in
+  let dump = Scheduler.Dump.make () in
+  assert_bool "dump_contains_item" (List.exists (fun x -> x.Scheduler.Dump.thing = "test_dump") dump)
+
+let _ = Scheduler.start ()
+
+let tests =
```

Comment:
Or maybe just add a small explanation in the commit message of d60dea6 would be the best